application/Core/MORR/Data/Transcoding/MPEG/MPEGEncoder.cs

```diff
@@ -0,0 +1,198 @@
+﻿using System;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Windows.Media.Core;
+using Windows.Media.MediaProperties;
+using Windows.Media.Transcoding;
+using MORR.Core.Data.Capture.Video;
+using MORR.Core.Data.Transcoding.Exceptions;
+using MORR.Shared.Events.Queue;
+using MORR.Shared.Utility;
+
+namespace MORR.Core.Data.Transcoding.MPEG
+{
+    public class MPEGEncoder : IEncoder
+    {
+        private readonly ManualResetEvent onResolutionInferred = new ManualResetEvent(false);
+        private readonly AutoResetEvent nextSampleReady = new AutoResetEvent(false);
+        private readonly AutoResetEvent sampleProcessed = new AutoResetEvent(true);
+        private DateTime encodingStart;
+        private DirectXVideoSample? nextSample;
+        private Tuple<uint, uint>? inferredResolution;
+        private bool hasInferredResolution;
+
+        [Import]
+        private MPEGEncoderConfiguration Configuration { get; set; }
+
+        [Import]
+        private IEncodeableEventQueue<DirectXVideoSample> VideoQueue { get; set; }
+
+        public void Encode(DirectoryPath recordingDirectoryPath)
+        {
+            encodingStart = DateTime.Now;
+            var recordingFilePath = GetRecordingFile(recordingDirectoryPath);
+
+            Task.Run(ConsumeVideoSamples);
+            Task.Run(() => InitializeTranscode(recordingFilePath));
+        }
+
+        private FilePath GetRecordingFile(DirectoryPath recordingDirectoryPath)
+        {
+            var recordingFilePath =
+                Path.Combine(recordingDirectoryPath.ToString(), $"{Configuration.RecordingName}.mp4");
```

Comment:
I get your point and I agree that that may be useful information to have, however, is this really the place to do it or should we rather change how the session manager creates a new recording directory? I currently imagine our final output will be a directory with two files: One .mp4 file containing the desktop capture and one .json file containing the event data. We could now either have both `MPEGEncoder` and `JsonEncoder` name their files by including the timestamp or we could just include the timestamp in the recording directory. I have no real preference, but I would imagine that getting the data back when only the recording directory is named could be slightly more annoying - what do you think?

Reply:
> [...]but I would imagine that getting the data back when only the recording directory is named could be slightly more annoying

I think i don't really get what you mean, do you refer to the filenames being less predictable? It would also be fine to add the timestamp information to the directory (their names would become pretty long though). We could also switch it around and give the directories a timestamp and the GUID to the filename, it doesn't really matter i guess.

I however agree that it shouldn't be done here where i commented it, but when building the directorypath/recordingname (deciding this in the invidual encoders would indeed be bad).

Reply:
My concern is that linking the timestamp to a file via its parent directory would disallow disassociating these files from their directory, which may be a first step in analysis (what I am trying to say is that I suspect that everything but the actual event data json files will get thrown away almost immediately, and if the files themselves don't contain the timestamp, this process would involve additional processing to get the timestamp from the parent directory and add it to the file; though the file currently isn't uniquely named anyway, so this is not a straightforward process).
I don't think we should split the GUID/timestamp part, because both apply to the same scope (either one timestamp and one GUID per recording or one timestamp and one GUID per file, though they should probably be the same among all files in that directory).
I only really see two options: Having the timestamp/GUID in the recording directory or having the timestamp/GUID in the invidivual file (and thus probably duplicated and encoder-specific). I dislike the first option because it makes analysis more complicated (see above) and I dislike the second option because it adds additonal logic to the encoders and duplicates information among them, bringing us slightly further away from the specified single-file design (and making the specific file names be less about the content that they hold in relation to the entire directory and more about additional metadata that would also be relevant for other files in the same directory).

Reply:
I am resolving this for now - we can carry on with this discussion when it comes to implementing the `JsonEncoder`, which will require further changes to Core anyway